lib/SILOptimizer/Analysis/SimplifyInstruction.cpp

```diff
@@ -727,45 +746,556 @@ SILValue swift::simplifyInstruction(SILInstruction *I) {
+  SingleValueInstruction *oldValue;
+  SILValue newValue;
+  std::function<void(SILInstruction *)> eraseNotify;
+  std::function<void(SILInstruction *)> newInstNotify;
+  DeadEndBlocks *deBlocks = nullptr;
+
+  SILBasicBlock::iterator perform();
+
+  ValueOwnershipKindPair getClassifierPair() const {
+    return getValueOwnershipKindPair(SILValue(oldValue).getOwnershipKind(),
+                                     newValue.getOwnershipKind());
+  }
+
+  /// Lifetime extend newValue over owned oldValue assuming that our copy will
+  /// be consumed by oldValue's consuming uses after RAUWing.
+  SILValue copyAndExtendForConsumingRAUW();
+
+  /// Lifetime extend newValue over the value oldValue in preparation for
+  /// RAUWing assuming that oldValue is not owned and thus we must lifetime
+  /// extend the given value over all uses.
+  SILValue copyAndExtendForNonConsumingRAUW();
+
+  /// Lifetime extend newValue over the value \p oldValue in preparation for
+  /// RAUWing assuming that oldValue is not owned and thus we must lifetime
+  /// extend over all of oldValue's uses.
+  SILValue copyBorrowAndExtendForRAUW();
+
+  SILBasicBlock::iterator handleReplaceOfNone();
+
+  OwnershipLifetimeExtender getLifetimeExtender() {
+    return {eraseNotify, newInstNotify, deBlocks};
+  }
+};
+
+} // anonymous namespace
+
+SILValue OwnershipRAUWUtility::copyAndExtendForConsumingRAUW() {
+  return getLifetimeExtender().copyAndExtendForConsumingRAUW(newValue, oldValue);
+}
+
+SILValue OwnershipRAUWUtility::copyAndExtendForNonConsumingRAUW() {
+  return getLifetimeExtender().copyAndExtendForNonConsumingRAUW(newValue, oldValue->getUses());
+}
+
+// LHS is owned value, RHS is borrow.
+SILValue OwnershipRAUWUtility::copyBorrowAndExtendForRAUW() {
+  return getLifetimeExtender().copyBorrowAndExtendForRAUW(newValue, oldValue->getUses());
+}
+
+static void cleanupOperandsBeforeDeletion(
+    SILInstruction *oldValue,
+    std::function<void(SILInstruction *)> newNotifyInst) {
+  SILBuilderWithScope builder(oldValue);
+  for (auto &op : oldValue->getAllOperands()) {
+    if (!op.isConsumingUse()) {
+      continue;
+    }
+
+    switch (op.get().getOwnershipKind()) {
+    case ValueOwnershipKind::Owned: {
+      auto *dvi = builder.createDestroyValue(oldValue->getLoc(), op.get());
+      if (newNotifyInst)
+        newNotifyInst(dvi);
+      continue;
+    }
+    case ValueOwnershipKind::Guaranteed: {
+      // Should only happen once we model destructures as true reborrows.
+      auto *ebi = builder.createEndBorrow(oldValue->getLoc(), op.get());
+      if (newNotifyInst)
+        newNotifyInst(ebi);
+      continue;
+    }
+    case ValueOwnershipKind::None:
+      continue;
+    case ValueOwnershipKind::Unowned:
+      llvm_unreachable("Unowned object can never be consumed?!");
+    }
+    llvm_unreachable("Covered switch isn't covered");
+  }
+}
+
+static ValueOwnershipKind findCompatibleNonNoneOwnershipKind(OperandOwnershipKindMap map) {
```

Comment:
I'm worried to see new code dealing with `OwnershipKindMaps` instead of a simpler lifetime constraint model. The kind maps severely obscure the ownership model and I'm afraid it will be more difficult to eliminate the kind maps later if they keep getting propagated through APIs instead of being encapsulated behind a simple interface for "ownership constraints" ('owned', 'guaranteed', 'any') plus "lifetime constraints" ('consuming', 'immediate', 'scoped', 'propagating/escaping').

In this case, the entire `findCompatibleNonNoneOwnershipKind` helper should just be a trivial lattice 'join' over the new ownership kind and the operand's ownership constraint.